File: character_os/simulation/runner.py
"""Run repeated Character Physics events and record snapshots."""

from __future__ import annotations

from dataclasses import dataclass

from memory.decay import decay_memories, effective_memory_weight
from physics_engine import (
    CharacterPhysicsEngine,
    CharacterPhysicsState,
    ExperienceEvent,
    PhysicsStepResult,
)


@dataclass(frozen=True)
class SimulationSnapshot:
    """A compact snapshot after one event step."""

    step: int
    event_id: str
    memory_id: str
    category: str
    impact_score: float
    memory_repetition_count: int
    memory_recency: float
    memory_effective_weight: float
    cluster_mass: float
    cluster_density: float
    cluster_stability: float
    cluster_age: int
    coordinate: dict[str, float]


@dataclass(frozen=True)
class SimulationResult:
    """Result of a repeated-event simulation."""

    snapshots: list[SimulationSnapshot]
    final_state: CharacterPhysicsState


def run_event_sequence(
    state: CharacterPhysicsState,
    events: list[ExperienceEvent],
    engine: CharacterPhysicsEngine | None = None,
    days_per_step: float = 0.0,
) -> SimulationResult:
    """Process events in order and record personality-space snapshots."""

    engine = engine or CharacterPhysicsEngine()
    snapshots: list[SimulationSnapshot] = []

    for index, event in enumerate(events, start=1):
        if days_per_step > 0 and state.memories:
            state.memories = decay_memories(state.memories, days_per_step)
        step = engine.process_event(state, event)
        snapshots.append(_snapshot(index, step, state))

    return SimulationResult(snapshots=snapshots, final_state=state)


def _snapshot(
    step_index: int,
    step: PhysicsStepResult,
    state: CharacterPhysicsState,
) -> SimulationSnapshot:
    """Create one simulation snapshot."""

    _ = state
    cluster = step.cluster
    return SimulationSnapshot(
        step=step_index,
        event_id=step.event.id,
        memory_id=step.memory_node.id,
        category=step.particle.category,
        impact_score=step.impact_score.value,
        memory_repetition_count=step.memory_node.repetition_count,
        memory_recency=step.memory_node.recency,
        memory_effective_weight=effective_memory_weight(step.memory_node),
        cluster_mass=cluster.mass,
        cluster_density=cluster.density,
        cluster_stability=cluster.stability,
        cluster_age=cluster.age,
        coordinate=step.coordinate_drift.after.to_dict(),
    )

File: src/core/simulation/runner.ts
import { effectiveMemoryWeight, decayMemories } from "../memory/decay";
import { syncClustersWithGalaxyMetrics } from "../galaxy/clusterMetrics";
import type { BoundaryImpactResult } from "../boundary/psychologicalBoundary";
import {
  CharacterPhysicsEngine,
  type CharacterPhysicsState,
  type PhysicsStepResult
} from "../physics/physicsEngine";
import type { ExperienceEvent } from "../event/event";
import type { PersonalityCoordinateValues } from "../personality/coordinate";
import { coordinateToRecord } from "../personality/coordinate";
import { toGalaxyStepTrace, type GalaxyStepTrace } from "../trace/galaxyTrace";
import type { EventTemporalTrace } from "../time/eventTemporalSemantics";
import { UNKNOWN_EVENT_OCCURRED_AT } from "../time/eventTemporalSemantics";

export interface SimulationSnapshot {
  step: number;
  eventId: string;
  memoryId: string;
  category: string;
  impactScore: number;
  memoryRepetitionCount: number;
  memoryRecency: number;
  memoryEffectiveWeight: number;
  clusterMass: number;
  clusterDensity: number;
  clusterStability: number;
  clusterAge: number;
  force: PersonalityCoordinateValues;
  velocity: PersonalityCoordinateValues;
  coordinate: PersonalityCoordinateValues;
  boundaryImpact: BoundaryImpactResult;
  galaxyTrace: GalaxyStepTrace;
  temporalSemantics: EventTemporalTrace;
}

export interface SimulationResult {
  snapshots: SimulationSnapshot[];
  finalState: CharacterPhysicsState;
}

export function runEventSequence(params: {
  state: CharacterPhysicsState;
  events: ExperienceEvent[];
  engine?: CharacterPhysicsEngine;
  daysPerStep?: number;
}): SimulationResult {
  const engine = params.engine ?? new CharacterPhysicsEngine();
  const snapshots: SimulationSnapshot[] = [];

  params.events.forEach((event, index) => {
    const hasExplicitEventTime = Boolean(
      event.occurredAt &&
      event.occurredAt !== UNKNOWN_EVENT_OCCURRED_AT &&
      Number.isFinite(Date.parse(event.occurredAt)),
    );
    // daysPerStep is a legacy memory-only simulation control. Explicit event
    // timestamps own elapsed-time recovery and must not be decayed twice.
    if ((params.daysPerStep ?? 0) > 0 && params.state.memories.length > 0 && !hasExplicitEventTime) {
      params.state.memories = decayMemories(params.state.memories, params.daysPerStep ?? 0);
      params.state.clusters = syncClustersWithGalaxyMetrics(params.state.clusters, params.state.memories);
    }
    const step = engine.processEvent(params.state, event);
    snapshots.push(snapshot(index + 1, step, params.state));
  });

  return { snapshots, finalState: params.state };
}

function snapshot(
  stepIndex: number,
  step: PhysicsStepResult,
  state: CharacterPhysicsState
): SimulationSnapshot {
  return {
    step: stepIndex,
    eventId: step.event.id,
    memoryId: step.memoryNode.id,
    category: step.particle.category,
    impactScore: step.impactScore.value,
    memoryRepetitionCount: step.memoryNode.repetitionCount,
    memoryRecency: step.memoryNode.recency,
    memoryEffectiveWeight: effectiveMemoryWeight(step.memoryNode),
    clusterMass: step.cluster.mass,
    clusterDensity: step.cluster.density,
    clusterStability: step.cluster.stability,
    clusterAge: step.cluster.age,
    force: coordinateToRecord(step.coordinateDrift.totalForce),
    velocity: coordinateToRecord(state.velocity),
    coordinate: coordinateToRecord(step.coordinateDrift.after),
    boundaryImpact: step.boundaryImpact,
    galaxyTrace: toGalaxyStepTrace(step),
    temporalSemantics: step.temporalSemantics,
  };
}
